File: examples/extensions/go/gebext.go
package main

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"net"
)

const (
	frameJSON   byte = 0
	frameBinary byte = 1
)

type extRequest struct {
	ID     int64          `json:"id"`
	Fn     string         `json:"fn"`
	Args   []any          `json:"args"`
	Kwargs map[string]any `json:"kwargs"`
	Slots  int            `json:"slots"`
}

type extHandler func(fn string, args []any, kwargs map[string]any, slots [][]byte) (any, [][]byte, error)

func writeFrame(w io.Writer, typ byte, payload []byte) error {
	header := make([]byte, 5)
	binary.BigEndian.PutUint32(header[:4], uint32(len(payload)))
	header[4] = typ
	if err := writeAll(w, header); err != nil {
		return err
	}
	return writeAll(w, payload)
}

func writeAll(w io.Writer, data []byte) error {
	for len(data) > 0 {
		n, err := w.Write(data)
		if err != nil {
			return err
		}
		if n == 0 {
			return io.ErrShortWrite
		}
		data = data[n:]
	}
	return nil
}

func readFrame(r io.Reader) (byte, []byte, error) {
	header := make([]byte, 5)
	if _, err := io.ReadFull(r, header); err != nil {
		return 0, nil, err
	}
	length := binary.BigEndian.Uint32(header[:4])
	payload := make([]byte, length)
	if _, err := io.ReadFull(r, payload); err != nil {
		return 0, nil, err
	}
	return header[4], payload, nil
}

func serveExtension(conn net.Conn, name string, functions []string, handler extHandler) {
	defer conn.Close()
	handshake, _ := json.Marshal(map[string]any{"v": 1, "name": name, "functions": functions})
	if err := writeFrame(conn, frameJSON, handshake); err != nil {
		return
	}
	for {
		typ, payload, err := readFrame(conn)
		if err != nil {
			return
		}
		if typ != frameJSON {
			continue
		}
		var req extRequest
		if err := json.Unmarshal(payload, &req); err != nil {
			continue
		}
		slots := make([][]byte, req.Slots)
		for i := range slots {
			slotType, slot, err := readFrame(conn)
			if err != nil || slotType != frameBinary {
				return
			}
			slots[i] = slot
		}
		if req.Fn == "__shutdown__" {
			return
		}
		value, outSlots, callErr := handler(req.Fn, req.Args, req.Kwargs, slots)
		resp := map[string]any{"id": req.ID, "ok": callErr == nil, "value": value}
		if callErr != nil {
			resp["error"] = callErr.Error()
		}
		if len(outSlots) > 0 {
			resp["slots"] = len(outSlots)
		}
		data, _ := json.Marshal(resp)
		_ = writeFrame(conn, frameJSON, data)
		for _, slot := range outSlots {
			_ = writeFrame(conn, frameBinary, slot)
		}
	}
}

func bytesArg(value any, slots [][]byte) ([]byte, error) {
	marker, ok := value.(map[string]any)
	if !ok || marker["$type"] != "bytes" {
		return nil, fmt.Errorf("expected bytes marker")
	}
	slot := int(marker["slot"].(float64))
	if slot < 0 || slot >= len(slots) {
		return nil, fmt.Errorf("bytes slot out of range")
	}
	return slots[slot], nil
}

File: examples/extensions/node/gebext.js
const FRAME_JSON = 0;
const FRAME_BINARY = 1;

function frame(type, payload = Buffer.alloc(0)) {
  const header = Buffer.alloc(5);
  header.writeUInt32BE(payload.length, 0);
  header.writeUInt8(type, 4);
  return Buffer.concat([header, payload]);
}

function makeReader(socket) {
  let buffer = Buffer.alloc(0);
  const waiters = [];
  socket.on("data", chunk => {
    buffer = Buffer.concat([buffer, chunk]);
    while (waiters.length) waiters.shift()();
  });
  socket.on("close", () => {
    while (waiters.length) waiters.shift()();
  });
  return async function readFrame() {
    while (buffer.length < 5) {
      if (socket.destroyed) return null;
      await new Promise(resolve => waiters.push(resolve));
    }
    const length = buffer.readUInt32BE(0);
    const type = buffer.readUInt8(4);
    while (buffer.length < 5 + length) {
      if (socket.destroyed) return null;
      await new Promise(resolve => waiters.push(resolve));
    }
    const payload = buffer.subarray(5, 5 + length);
    buffer = buffer.subarray(5 + length);
    return {type, payload};
  };
}

function decode(value, slots) {
  if (value && value.$type === "bytes") return slots[value.slot];
  return value;
}

function encode(value, slots) {
  if (Buffer.isBuffer(value)) {
    const slot = slots.length;
    slots.push(value);
    return {$type: "bytes", slot};
  }
  return value;
}

async function serveExtension(socket, name, functions, handler) {
  socket.write(frame(FRAME_JSON, Buffer.from(JSON.stringify({
    v: 1,
    name,
    functions: [...functions].sort(),
  }))));
  const readFrame = makeReader(socket);
  while (!socket.destroyed) {
    const first = await readFrame();
    if (!first) return;
    if (first.type !== FRAME_JSON) continue;
    const req = JSON.parse(first.payload.toString("utf8"));
    const slots = [];
    for (let i = 0; i < (req.slots || 0); i++) {
      const slot = await readFrame();
      if (!slot || slot.type !== FRAME_BINARY) throw new Error("expected binary slot");
      slots.push(slot.payload);
    }
    if (req.fn === "__shutdown__") return socket.end();
    try {
      const args = (req.args || []).map(arg => decode(arg, slots));
      const kwargs = {};
      for (const [key, value] of Object.entries(req.kwargs || {})) {
        kwargs[key] = decode(value, slots);
      }
      const result = await handler(req.fn, args, kwargs);
      const outSlots = [];
      const value = encode(result, outSlots);
      const resp = {id: req.id, ok: true, value};
      if (outSlots.length) resp.slots = outSlots.length;
      socket.write(frame(FRAME_JSON, Buffer.from(JSON.stringify(resp))));
      for (const slot of outSlots) socket.write(frame(FRAME_BINARY, slot));
    } catch (error) {
      socket.write(frame(FRAME_JSON, Buffer.from(JSON.stringify({
        id: req.id || 0,
        ok: false,
        error: error.message,
      }))));
    }
  }
}

module.exports = {FRAME_JSON, FRAME_BINARY, frame, makeReader, decode, encode, serveExtension};
